### PR TITLE
Enable SAPSystems projection

### PR DESCRIPTION
**Pull request by dottorblaster**
This PR is a first step toward a makeover of the SAP Systems view. With this we can project the raw data onto entities persisted inside the DB, for further aggregation.

### CODE REVIEW

File: web/entities/sapsystem.go [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package entities
+
+import (
+	"time"
+)
+
+type SAPSystemInstance struct {
```
Comment (arbulu89) [inline web/entities/sapsystem.go:7]: I thinks we would need to have the `SAPSystem` and `SAPInstance` structs combo.
Why?
Because we can have some SAP instance with a different SID than the SAP system itself (the database doesn't need to have the same SID than the SAP system).

Later on, 1 SAP system is composed by multiple instances, that can have different SID and type. And even different parameters depending on the type. For example, as `application` type instances don't have any System replication data.

Maybe some kind of polymorfism would be a good idea here.

**Review by arbulu89 — COMMENTED** [review]
Hey @dottorblaster ,
I'm afraid this implementation can be too simplistic. We would need to see better how the SAP systems composition is done to relate properly the systems with the databases, and all the instances within

File: web/entities/sapsystem.go (hunk 1 above)
Comment (fabriziosestito) [reply web/entities/sapsystem.go:7]: @arbulu89 the idea is that this is how we store the data from the events, the `SAPSystem`, aggregation would come in the service when we will transform (group) instances to a `model.SAPSystem` which will hold many instances.
The database aggregation is quite easy since we will search for a `entities.Host` having the same IP as the `DBHost` field here.
As for the SystemReplication data or other fields the idea was to keep them empty in case of the `application` type and just ignore them when casting that to a model.
Basically, polymorphism will be handled at service level not at persistence level.

Comment (dottorblaster) [reply web/entities/sapsystem.go:7]: @fabriziosestito said that way better than me :smile: 

I agree that we need to tell the difference between different models, but this shall be done at service level, this PR only introduces a basic projector so we can move forward with other works

Comment (arbulu89) [reply web/entities/sapsystem.go:7]: @fabriziosestito Ok, fair enough. I don't think it will scale well in the future, because we will need to have SAP system specific data, but I guess we will be able to change it

File: web/datapipeline/sap_systems_projector.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	"fmt"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/trento-project/trento/internal/sapsystem"
+	"github.com/trento-project/trento/web/entities"
+	"gorm.io/gorm"
+	"gorm.io/gorm/clause"
+)
+
+const SAPSystemsApplication = "application"
```
Comment (arbulu89) [inline web/datapipeline/sap_systems_projector.go:13]: I think we could unify these names in `internal` and use the same value in the agent and here. We can even change the agent to send a string if it helps.

File: web/datapipeline/sap_systems_projector.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+func SAPSystemsProjector_SAPSystemsDiscoveryHandler(dataCollectedEvent *DataCollectedEvent, db *gorm.DB) error {
+	decoder := getPayloadDecoder(dataCollectedEvent.Payload)
+	var discoveredSAPSystems sapsystem.SAPSystemsList
+	if err := decoder.Decode(&discoveredSAPSystems); err != nil {
+		log.Errorf("can't decode data: %s", err)
+		return err
+	}
+
+	for _, s := range discoveredSAPSystems {
```
Comment (arbulu89) [inline web/datapipeline/sap_systems_projector.go:32]: I see some issues in this loop. 
Here, we are looping the SAP systems installed in 1 host. There might be 1..n system.
But within the SAP system, we can have multiple SAP instances installed, 1..n again. So, I miss an additional loop here, which creates the SAP instance entry itself.

Example: We have a netweaver installation that has the ASCS and PAS in the same machine. Both belong to the same SAP system, but are 2 different instance (2 rows in the UI table).

So most probably, we would need to run the `storeSAPInstance` function within the `s.Instaces` loop.

That's why I see that the instances and systems are kind of coupled. But I guess we can introduce the system data in the instance as well (as they are parent/children somehow)

**Review by arbulu89 — CHANGES_REQUESTED** [review]
Some more comments here

File: web/datapipeline/sap_systems_projector.go (hunk 3 above)
Comment (dottorblaster) [reply web/datapipeline/sap_systems_projector.go:32]: > That's why I see that the instances and systems are kind of coupled

Absolutely, I see the point but I wanted to do this at service level, populating a `SAPSystem` model grouping many `SAPSystemsInstance` entities cast into proper models.

> Example: We have a netweaver installation that has the ASCS and PAS in the same machine. Both belong to the same SAP system, but are 2 different instance (2 rows in the UI table).

Understood, I'll try to come up with something better, thanks for the feedback

File: web/datapipeline/sap_systems_projector.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+		instance := entities.SAPSystemInstance{
+			AgentID:  dataCollectedEvent.AgentID,
+			SystemID: s.Id,
+			SID:      s.SID,
+		}
+
+		if s.Type == 1 {
```
Comment (arbulu89) [inline web/datapipeline/sap_systems_projector.go:36]: Could we put this in a `switch/cas` fashion at least. It looks the most appropriate way

**Review by fabriziosestito — APPROVED** [review]
LGTM! Good first iteration

File: web/datapipeline/sap_systems_projector.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+	var discoveredSAPSystems sapsystem.SAPSystemsList
+	if err := decoder.Decode(&discoveredSAPSystems); err != nil {
+		log.Errorf("can't decode data: %s", err)
+		return err
+	}
+
+	for _, s := range discoveredSAPSystems {
+		instance := entities.SAPSystemInstance{
+			AgentID:  dataCollectedEvent.AgentID,
+			SystemID: s.Id,
```
Comment (arbulu89) [inline web/datapipeline/sap_systems_projector.go:32]: I'm a bit troubled with the name, but I recognize that it is not simple hehe
Maybe, we could rename `SystemID` to simply `ID`? `SID` at the end is `system id` as well
It is the internal ID of the system at the end.
By the way, in the future we will need to convert this to `uuid` format. Nothing to be done in the this PR though

File: web/datapipeline/sap_systems_projector.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+		}
+
+		for _, i := range s.Instances {
+			var features string
+			var instanceNumber string
+
+			if p, ok := i.SAPControl.Properties["SAPSYSTEM"]; ok {
+				instanceNumber = p.Value
+			} else {
+				continue
```
Comment (arbulu89) [inline web/datapipeline/sap_systems_projector.go:52]: This `else` case is something that shouldn't never happen. Could we put some kind of `warning` logging?

File: web/datapipeline/sap_systems_projector.go [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+			}
+
+			for _, SAPControlInstance := range i.SAPControl.Instances {
+				if instanceNumber == fmt.Sprintf("%02d", SAPControlInstance.InstanceNr) {
+					features = SAPControlInstance.Features
```
Comment (arbulu89) [inline web/datapipeline/sap_systems_projector.go:57]: We can exit from the loop once this condition is met, as we won't never have it as true again for this sap control instance, or instance in general. Let's add a break

**Review by nelsonkopliku — APPROVED** [review]
Very nice! Step ahead! :rocket:

File: web/datapipeline/sap_systems_projector.go [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+				}
+			}
+
+			instance.Features = features
+			instance.InstanceNumber = instanceNumber
+
+			err := storeSAPInstance(db, instance, "system_id", "sid", "type", "description", "features", "instance_number")
```
Comment (arbulu89) [inline web/datapipeline/sap_systems_projector.go:64]: I would vote to **not** update the `description` as it is empty always. In fact, this field won't be auto-discovered. It is an old discussion, but it doesn't make much sense by now hehe

**Review by arbulu89 — CHANGES_REQUESTED** [review]
Looking good!
I have added some more comments. Most of them are just my personal thought, feel free to not apply them if needed, but I think they deserve a little look

File: web/datapipeline/sap_systems_projector.go (hunk 5 above)
Comment (fabriziosestito) [reply web/datapipeline/sap_systems_projector.go:32]: @arbulu89 I asked @dottorblaster  to change this 🍡 
To be aligned, I see this expressed as SystemID in the UI, what is the SAP domain word for this one?

Comment (dottorblaster) [reply web/datapipeline/sap_systems_projector.go:32]: Tough decision but `ID` is a slippery call :sweat_smile: I would prefer having little longer field names, and `ID` is just... well an ID? :smile:

File: web/datapipeline/sap_systems_projector.go (hunk 7 above)
Comment (dottorblaster) [reply web/datapipeline/sap_systems_projector.go:57]: I don't like breaks, I do this only because I love you

File: web/datapipeline/sap_systems_projector.go (hunk 5 above)
Comment (arbulu89) [reply web/datapipeline/sap_systems_projector.go:32]: The thing is that `SystemID` and `SID` are the same as `SID` stands for `System identifier`. The thing is that our `SystemID` is a an internal auto-generated ID that is **global**. The sap `SID` never was created to be global, that's why it is just a 3 chars/digits string.
In this case the `ID` is our ID, the identifier of the SAP system (not the instance).
Up to you guys, I just wanted to comment that `SystemID` and `SID` are synonims hehe One is used by SAP and the other by us

File: web/datapipeline/sap_systems_projector.go [hunk 9]
```diff
@@ -49,19 +50,21 @@ func SAPSystemsProjector_SAPSystemsDiscoveryHandler(dataCollectedEvent *DataColl
 			if p, ok := i.SAPControl.Properties["SAPSYSTEM"]; ok {
 				instanceNumber = p.Value
 			} else {
+				log.Warn("Instance Number not found")
```
Comment (arbulu89) [inline web/datapipeline/sap_systems_projector.go:53]: Let's improve this to be `log.Warnf("Instance Number not found for %s", s.SID)`

**Review by arbulu89 — APPROVED** [review]
Thanks @dottorblaster ,
Just added other 2 small comments XD but we don't need to have other round of review hehe

File: web/datapipeline/sap_systems_projector.go (hunk 5 above)
Comment (fabriziosestito) [reply web/datapipeline/sap_systems_projector.go:32]: Ah alright i thought that `SystemID` was our `ID` and `SID` was another thing :D

Comment (arbulu89) [reply web/datapipeline/sap_systems_projector.go:32]: Yes, and it is.
`SystemID` our internal ID
`SID` sap system id -> `System identifier`
The difference is that one is ours and the other SAP internal hehe
